src/aws/mod.rs

```diff
@@ -202,166 +200,92 @@ pub async fn sign_request(
-}
-
-impl<S> Layer<S> for CaptureRequestSize {
-    type Service = CaptureRequestSizeService<S>;
-
-    fn layer(&self, inner: S) -> Self::Service {
-        CaptureRequestSizeService {
-            enabled: self.enabled,
+impl<T> aws_smithy_runtime_api::client::http::HttpClient for AwsHttpClient<T>
+where
+    T: aws_smithy_runtime_api::client::http::HttpClient,
+{
+    fn http_connector(
+        &self,
+        settings: &aws_smithy_runtime_api::client::http::HttpConnectorSettings,
+        components: &aws_sdk_cloudwatch::config::RuntimeComponents,
+    ) -> SharedHttpConnector {
+        let http_connector = self.http.http_connector(settings, components);
+
+        SharedHttpConnector::new(AwsConnector {
             region: self.region.clone(),
-            inner,
-        }
+            http: http_connector,
+        })
     }
 }
 
-/// Service for capturing the payload size for AWS API client requests and emitting internal telemetry.
-#[derive(Clone)]
-struct CaptureRequestSizeService<S> {
-    enabled: bool,
+#[derive(Clone, Debug)]
+pub struct AwsConnector<T> {
+    http: T,
     region: Region,
-    inner: S,
 }
 
-impl<S> Service<Request> for CaptureRequestSizeService<S>
+impl<T> HttpConnector for AwsConnector<T>
 where
-    S: Service<Request, Response = Response> + Send + Sync + 'static,
-    S::Future: Send + 'static,
+    T: HttpConnector,
 {
-    type Response = S::Response;
-    type Error = S::Error;
-    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
-
-    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        self.inner.poll_ready(cx)
-    }
-
-    fn call(&mut self, req: Request) -> Self::Future {
-        // Attach a body callback that will capture the bytes sent by interrogating the body chunks that get read as it
-        // sends the request out over the wire. We'll read the shared atomic counter, which will contain the number of
-        // bytes "read", aka the bytes it actually sent, if and only if we get back a successful response.
-        let (req, maybe_bytes_sent) = if self.enabled {
-            let shared_bytes_sent = Arc::new(AtomicUsize::new(0));
-            let (request, properties) = req.into_parts();
-            let (parts, body) = request.into_parts();
-
-            let body = {
-                let shared_bytes_sent = Arc::clone(&shared_bytes_sent);
-
-                body.map_immutable(move |body| {
-                    let body = MeasuredBody::new(body, Arc::clone(&shared_bytes_sent));
-                    SdkBody::from_dyn(BoxBody::new(body))
-                })
-            };
-
-            let req = Request::from_parts(http::Request::from_parts(parts, body), properties);
-
-            (req, Some(shared_bytes_sent))
-        } else {
-            (req, None)
-        };
-
+    fn call(&self, req: HttpRequest) -> HttpConnectorFuture {
+        let byte_size = req.body().bytes().expect("body cannot be lazy").len();
```

Comment:
This will panic if we start sending lazy http bodies to AWS.  If 99% sure we never do this, and if we ever did the panic should be caught in testing, so all should be good. 

The previous version wrapped the http call in a tower layer so it was possible to inject our own http `Body` that counted the bytes even with a lazy body. I can't get this to work with this method due to type system awkwardness.

I'm not totally happy with it though, so if anyone knows a better way around this I'm all ears.

Reply:
Could we use a similar approach here to what we had before, adding a map over the body?

Something like the below (does not compile)
```
    let shared_bytes_sent = Arc::new(AtomicUsize::new(0));
    let mut parts = req.into_parts();
    let body = parts.body.map_preserve_contents(move |body| {
        let body = MeasuredBody::new(body, Arc::clone(&shared_bytes_sent));
        SdkBody::from_dyn(BoxBody::new(body))
    });
```

Reply:
By jove, yes that seems to work.